packages/dds/quorum/src/quorum.ts

```diff
@@ -0,0 +1,373 @@
+        this.submitLocalMessage(noOp);
+    };
+
+    private readonly handleIncomingNoOp = (minimumSequenceNumber: number) => {
+        // Run through each of the values, find any pending that should now be considered settled because the MSN has
+        // passed when they were sequenced.
+        for (const [key, value] of this.values) {
+            const pending = value.pending;
+            if (pending !== undefined && minimumSequenceNumber >= pending.sequenceNumber) {
+                // The pending value has settled
+                if (pending.type === "set") {
+                    this.values.set(key, {
+                        accepted: { value: pending.value, sequenceNumber: minimumSequenceNumber },
+                        pending: undefined,
+                    });
+                } else if (pending.type === "delete") {
+                    this.values.delete(key);
+                }
+                this.emit("accepted", key);
+            }
+        }
+    };
+
+    /**
+     * Create a summary for the quorum
+     *
+     * @returns the summary of the current state of the quorum
+     * @internal
+     */
+    protected summarizeCore(serializer: IFluidSerializer): ISummaryTreeWithStats {
+        const content = [...this.values.entries()];
+        return createSingleBlobSummary(snapshotFileName, JSON.stringify(content));
+    }
+
+    /**
+     * {@inheritDoc @fluidframework/shared-object-base#SharedObject.loadCore}
+     * @internal
+     */
+    protected async loadCore(storage: IChannelStorageService): Promise<void> {
+        const content = await readAndParse<[string, QuorumValue][]>(storage, snapshotFileName);
+        content.forEach(([key, value]) => {
+            this.values.set(key, value);
+        });
+    }
+
+    /**
+     * @internal
+     */
+    protected initializeLocalCore() { }
+
+    /**
+     * @internal
+     */
+    protected onDisconnect() {
+        this.disconnectWatcher.emit("disconnect");
+    }
+
+    /**
+     * Override resubmit core to avoid resubmission on reconnect.
+     * TODO It's probably ok to resubmit these?
+     * @internal
+     */
+    protected reSubmitCore() { }
+
+    /**
+     * Process a quorum operation
+     *
+     * @param message - the message to prepare
+     * @param local - whether the message was sent by the local client
+     * @param localOpMetadata - For local client messages, this is the metadata that was submitted with the message.
+     * For messages from a remote client, this will be undefined.
+     * @internal
+     */
+    protected processCore(message: ISequencedDocumentMessage, local: boolean, localOpMetadata: unknown) {
+        if (message.type === MessageType.Operation) {
+            const op = message.contents as IQuorumOperation;
+
+            switch (op.type) {
+                case "set":
+                    this.incomingOp.emit("set", op.key, op.value, op.refSeq, message.sequenceNumber);
```

Comment:
I'm not sure I see the value of having this.incomingOp, given that there is exactly one consumer - why not directly call this.handleIncomingSet() from here?

Reply:
Same thought on this.disconnectWatcher.

Reply:
Good question -- I don't take advantage of this yet in `Quorum`, but it becomes useful for resolving/rejecting promise-based APIs in particular.  I used this pattern in `TaskManager` for a better example of it in use today, and I expect I'll be considering promise-based APIs in `Quorum` in a future iteration.

The `EventEmitter` pattern makes it really easy to resolve/reject promises that are influenced by outside state without needing to use a `Deferred` or otherwise smuggle out the `resolve()`/`reject()` methods (e.g. in CRC it smuggles out a resolve in `localOpMetadata` to later be called in `processCore()`).  Instead of handing out access to the resolve/reject and calling them directly from within state-change handlers, we can instead have the state-change handlers emit an event that the Promise can observe and resolve/reject itself.

I focus on two main benefits from this:
1. The Promise's lifecycle becomes easier to understand, since the full set of transformations is contained within the scope of the Promise body.
2. It's not exactly RAII, but it can make it more-immediately-obvious when we should expect the promise's resources to be released and whether appropriate cleanup is being done.  As compared to:
    * "`BlobManager` doesn't `pendingBlobIds.clear()` on `"dispose"`, is that a leak/problem?"
    * "`DataStoreContexts.delete()` doesn't reject the `Deferred` context before deleting it, could this result in a hang?"

`lockAcquireP` in `TaskManager.lockTask()` is a good example.  Reading no more than the promise body, it's relatively easy to see three outside influences control the settling of the promise, and in what way.  We can have confidence the promise will get cleaned up after settling since we can see the listeners are all removed in these flows, and no references are leaked.

The implementation of `newAckBasedPromise()` is another good example of this pattern.  Its `rejectBecauseDispose()` is called in response to the runtime raising a `"dispose"` event, rather than needing to smuggle out its reject method to be called from some dispose handler.  Kind of ironically, the usage of `newAckBasedPromise()` encourages smuggling out the resolve, but this implementation avoids the smuggle.

So in the future, supposing that we want to return a `Promise<boolean>` from `set()` similar to CRC's `write()`, `incomingOp` will let us watch for the ack and resolve appropriately with true or false without needing to split the Promise's logic between `set()` and `processCore()` as CRC does -- it can live fully in `set()`.  Similar for any other state transition that we decide matters (disconnect/reconnect, dispose, etc.).

There are probably other non-`EventEmitter` patterns that offer similar benefits, but this is the first one I landed on that had the characteristics I was looking for.

Reply:
something that will be tricky with an promised based api will be dealing with minimumSequenceNumber which is global.

the second you go async other parts of the system can send and receive ops, which can then push msn forward. I think for this to work you'd need to pause the container. i haven't thought about it deeply, but there could be dead lock risks if there are multiple quourms

Reply:
There are two problems I see with asynchrony:
1. The one Tony mentions. The only solution there (that does not require global things that breaks composition / encapsulation) is ability of arbitrary scoped code to work with immutable document revisions (and be independent from the rest of the code flow). I think it's a must (in long term), I have a hard time seeing how complex systems can be built without it. It has its own issues (this works great for read-only flows, harder for read/write flows), various techniques exist to mitigate that.
2. Support for offline scenarios (area of active work) relies on ability to serialize and rehydrate full document state. It assumes that rehydrated container is a natural continuation of the same session (including how pending local changes / acks across those two sessions are handled). There is no way to serialize in-flight promises :) I believe this pushes us toward having more of our (DDS) APIs in the form of fire-and-forget + events. I.e. user code should be able to inspect document state (on load), "learn" about all the in-flight activity initiated by this client and re-establish all the right event handlers to continue as nothing happened. It's not a strict requirement (because app may chose to ignore such pending activity and treat it as "coming from some other client"), but I think there is limit to that.
   - I'm more and more convinced that we should reduce number of consensus-based flows (and DDSs). Basically "leader" election does not work (minutes to detect disconnects, ops generated by "leader" are submitted after it lost leadership, etc.). Most of such flows should instead rely on atomic completion of work. I.e. all clients racing to do work and first to submit result of the work wins, and that completes the work (does not start the work as in leader election flow). (Hurray to block chain :))
  
\# 2 could partially help with # 1. It's same thing (i.e. callbacks and promise are functionally the same, difference is only in syntactic sugar coating), but developers of await-heavy code have harder time reasoning about changing global state underneath due to asynchrony. It sucks (callback-based code), but I think that pushed in right direction - reduction of (the need) of these flows.